pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,226 @@
+	if err != nil {
+		return nil, err
+	}
+
+	zoneToInstanceTypeInfo := map[string][]*ec2.InstanceTypeInfo{}
+	for _, zone := range locations {
+		if instanceTypes, ok := p.instanceTypeCache.Get(zone); ok {
+			zoneToInstanceTypeInfo[zone] = instanceTypes.([]*ec2.InstanceTypeInfo)
+			continue
+		}
+
+		// populate the cache by zonal keys
+		instanceTypes, err := p.getAllInstanceTypes(ctx)
+		if err != nil {
+			return nil, err
+		}
+		instanceTypes, err = p.filterByLocation(ctx, instanceTypes, zone)
+		if err != nil {
+			return nil, err
+		}
+		p.instanceTypeCache.SetDefault(zone, instanceTypes)
+		zoneToInstanceTypeInfo[zone] = instanceTypes
+	}
+
+	for zone, instanceTypes := range zoneToInstanceTypeInfo {
+		zoneToInstanceTypeInfo[zone] = p.filterFrom(instanceTypes, constraints)
+	}
+	return zoneToInstanceTypeInfo, nil
+}
+
+// UniqueInstanceTypesFrom returns a unique slice of InstanceTypeInfo structs from the values of the instancePools passed in
+func (p *InstanceTypeProvider) UniqueInstanceTypesFrom(instancePools map[string][]*ec2.InstanceTypeInfo) []*ec2.InstanceTypeInfo {
+	result := []*ec2.InstanceTypeInfo{}
+	uniqueInstanceTypes := map[string]bool{}
+	for _, instanceTypes := range instancePools {
+		for _, instanceType := range instanceTypes {
+			if _, ok := uniqueInstanceTypes[*instanceType.InstanceType]; !ok {
+				uniqueInstanceTypes[*instanceType.InstanceType] = true
+				result = append(result, instanceType)
+			}
+		}
+	}
+	return result
+}
+
+// InstanceTypesPerZoneFrom returns a mapping of zone to InstanceTypeInfo based on the provided slice of instance types names and mapping of all zones to InstanceTypeInfo
+func (p *InstanceTypeProvider) InstanceTypesPerZoneFrom(instanceTypeNames []string, instancePools map[string][]*ec2.InstanceTypeInfo) map[string][]*ec2.InstanceTypeInfo {
+	result := map[string][]*ec2.InstanceTypeInfo{}
+	for zone, instanceTypes := range instancePools {
+		for _, instanceTypeName := range instanceTypeNames {
+			for _, instanceType := range instanceTypes {
+				if instanceType != nil && *instanceType.InstanceType == instanceTypeName {
+					result[zone] = append(result[zone], instanceType)
+				}
+			}
+		}
+	}
+	return result
+}
+
+// getAllInstanceTypes retrieves all instance types from the ec2 DescribeInstanceTypes API using some opinionated filters
+func (p *InstanceTypeProvider) getAllInstanceTypes(ctx context.Context) ([]*ec2.InstanceTypeInfo, error) {
+	instanceTypes := []*ec2.InstanceTypeInfo{}
+	describeInstanceTypesInput := &ec2.DescribeInstanceTypesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("supported-virtualization-type"),
+				Values: []*string{aws.String("hvm")},
+			},
+			{
+				Name:   aws.String("bare-metal"),
+				Values: []*string{aws.String("false")},
+			},
+			{
+				Name:   aws.String("instance-type"),
+				Values: aws.StringSlice([]string{"m5.*", "c5.*", "r5.*", "m5a.*", "c5a.*", "r5a.*", "m4.*", "c4.*", "r4.*", "m6g.*", "c6g.*", "r6g.*"}),
```

Comment:
Does this mean that we need customers to upgrade if EC2 releases new instance families?

Reply:
I would be concerned about this - I think we could exclude certain instance types, but I'd be wary of an inclusion list like this that needs to be updated.

Reply:
I loosened up the filters to whole families instead of generation (i..e `m*, c*, r*, etc`). They are also overridable by specifying instanceTypes on the node selector or in the instanceTypes provisioner crd.